feat: display environments list based on running pods

feloy opened the pull request:
### What does this PR do?

Lists the running environments, based on the running pods with appropriate labels.

For now, only the id of the recipe is displayed, more to come...

Only the operations on podman machines from the GUI are detected (see issue https://github.com/containers/podman-desktop/issues/5842)

### Screenshot / video of UI

<!-- If this PR is changing UI, please include 
screenshots or screencasts showing the difference -->

### What issues does this PR fix or reference?

Fixes #178, #179

### How to test this PR?

- pull an application for a recipe
- check the recipe ID is listed in the Environments page
- Stop or delete the pod created for the app from the GUI or CLI
- check the recipe ID disappears from the environments list
- stop te podman engine from the GUI
- check the recipe ID disappears from the environments list
- restart the podman machine (from GUI) and restart the pod (from GUI or CLI)
- check the recipe ID is listed in the Environments page
- reload the AI studio extension
- check the recipe ID is still listed in the Environments page
- restart Podman Desktop
- check the recipe ID is still listed in the Environments page

axel7083 commented on packages/backend/package.json:
```diff
@@ -38,6 +38,7 @@
     "watch": "vite --mode development build -w"
   },
   "dependencies": {
+    "@podman-desktop/api": "^0.0.202402061753-5c5d589",
```
will the feature in `0.0.202402061753-5c5d589` be in 1.8 ?

feloy replied:
@benoitf These 2 PRs will be needed by AI Studio in the next release:
- https://github.com/containers/podman-desktop/pull/5862
- https://github.com/containers/podman-desktop/pull/5864

Do I need to 'register' them somewhere, or will they be part of the next release automatically?

axel7083 replied:
@benoitf

benoitf replied:
automatic as it's in the main branch and 1.8 will use the main branch

benoitf replied:
but it means you can't cut new releases of AI studio using that API until 1.8 is out or you need to add some 'fallback' in the code

like if (api.engine.listPods) {...} else {....}

lstocchi replied:
you should update the entry in devDependencies or it will be duplicated

lstocchi commented on packages/backend/src/managers/environmentManager.ts:
```diff
@@ -0,0 +1,134 @@
+  adoptRunningEnvironments() {
+    this.podmanConnection.startupSubscribe(() => {
+      if (!containerEngine.listPods) {
+        // TODO(feloy) this check can be safely removed when podman desktop 1.8 is released
```
can we open an issue to keep track of this and the other similar TODO?

feloy replied:
right, created #242

lstocchi reviewed (commented):
I wonder why we don't show the pods when they are stopped. Shouldn't we be able to start an environment from this page later on?

feloy commented:
> I wonder why we don't show the pods when they are stopped. Shouldn't we be able to start an environment from this page later on?

I didn't want to complexify for now, by keeping only 2 states (running and non-running), and as restarting the pod is not in the list of features wwe want for now.

Also, as the start of the app container is managed by the extension (to restart it when the model is up and running), the user cannot simply restart the pod from podman desktop or the CLI, the extension would have to manage it, and that seems out of scope to me

benoitf commented on packages/backend/package.json:
```diff
@@ -38,11 +38,12 @@
     "watch": "vite --mode development build -w"
   },
   "dependencies": {
+    "@podman-desktop/api": "^0.0.202402061753-5c5d589",
```
FYI you should not have the API as a direct dependency (it's a dev dependency)

lstocchi commented:
> Also, as the start of the app container is managed by the extension (to restart it when the model is up and running), the user cannot simply restart the pod from podman desktop or the CLI, the extension would have to manage it, and that seems out of scope to me

I expect i can start an environment from this page otherwise what's the difference from the pod list page? It's just a filter.

I would expect that the environment store would keep a list of all environments found (running and stopped) then the svelte page filters it based on its needs. Maybe another page that uses this store would need to filter non-running or something else.

feloy commented:
> > Also, as the start of the app container is managed by the extension (to restart it when the model is up and running), the user cannot simply restart the pod from podman desktop or the CLI, the extension would have to manage it, and that seems out of scope to me
> 
> I expect i can start an environment from this page otherwise what's the difference from the pod list page? It's just a filter.
> 
> I would expect that the environment store would keep a list of all environments found (running and stopped) then the svelte page filters it based on its needs. Maybe another page that uses this store would need to filter non-running or something else.

I have added some status in the latest PR, to try and support the "stopped" pods. But there is a gray zone between deleted pod and running pod, where some containers are running only (status = Degraded in the pod), but I don't think we are getting enough events to correctly handle this case.

Please test and tell me what you think. I have added the status in the Environments page.

lstocchi commented:
> I have added some status in the latest PR, to try and support the "stopped" pods. But there is a gray zone between deleted pod and running pod, where some containers are running only (status = Degraded in the pod), but I don't think we are getting enough events to correctly handle this case.

Got it. Inside the `watchPod -> containerEngine.onEvent` method do we only receive the remove, start and stop state? If we receive other events we could update the environments store if there is some change in a pod state. 

> Please test and tell me what you think. I have added the status in the Environments page.

Great, thanks! 
The problem i have is that when i start a recipe, the pod state is running at first (bc all containers start fine), then the sample app fails to connect to the model service and the pod gets degraded, but as we just listen to the start, stop and delete events, i see the env in running state even if it should be 'starting' as the pod is degraded. But idk if we have enough info (see comment above) to fix it. If not, or you want to handle it later, we can remove the status from displaying, and re-add it on a follow-up PR. I also have to use this for my PR https://github.com/projectatomic/ai-studio/pull/223 and i can look at it.

lstocchi commented:
> So, I'll remove the latest commit (and keep it locally in another branch in case we need it later)

I would have preferred to just filter it in svelte but i don't want to block this even further.

jeffmaury commented on packages/frontend/src/lib/table/environment/ColumnName.svelte:
```diff
@@ -0,0 +1,9 @@
+</script>
+
+<div class="text-sm text-gray-700">
+  {object.recipeId}
```
Can't we display the state as well ?

feloy replied:
I'm waiting the design from Mo to work on the content here

feloy commented on packages/backend/src/managers/environmentManager.ts:
```diff
@@ -0,0 +1,134 @@
+
+    this.podmanConnection.onMachineStop(() => {
+      // Podman Machine has been stopped, we consider all recipe pods are stopped
+      this.#environments.clear();
```
How can I have several machines? On Mac, I'm able to start only one Podman machine at a given time

jeffmaury replied:
This is allowed on Windows

feloy replied:
I'll have a look on Windows and make a follow up PR if necessary. Thanks
